kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,442 @@
+/// > `delta.enableDeletionVectors` table property is not set.
+///
+/// See https://github.com/delta-io/delta/blob/master/PROTOCOL.md#deletion-vectors
+///
+/// TODO: When the kernel supports in-commit timestamps, we will also have to inspect CommitInfo
+/// actions to find the timestamp. These are generated when incommit timestamps is enabled.
+/// This must be done in the first phase because the second phase lazily transforms engine data with
+/// an extra timestamp column. Thus, the timestamp must be known ahead of time.
+/// See https://github.com/delta-io/delta-kernel-rs/issues/559
+///
+/// 2. Scan file generation phase [`LogReplayScanner::into_scan_batches`]: This performs another
+///    iteration over every action in the commit, and generates [`TableChangesScanData`]. It does
+///    so by transforming the actions using [`get_add_transform_expr`], and generating selection
+///    vectors with the following rules:
+///     - If a `cdc` action was found in the prepare phase, only `cdc` actions are selected
+///     - Otherwise, select `add` and `remove` actions. Note that only `remove` actions that do not
+///       share a path with an `add` action are selected.
+///
+/// Note: As a consequence of the two phases, LogReplayScanner will iterate over each action in the
+/// commit twice. It also may use an unbounded amount of memory, proportional to the number of
+/// `add` + `remove` actions in the _single_ commit.
+struct LogReplayScanner {
+    // True if a `cdc` action was found after running [`LogReplayScanner::prepare_table_changes`]
+    has_cdc_action: bool,
+    // A map from path to the deletion vector from the remove action. It is guaranteed that there
+    // is an add action with the same path in this commit
+    remove_dvs: HashMap<String, DvInfo>,
+    // The commit file that this replay scanner will operate on.
+    commit_file: ParsedLogPath,
+    // The timestamp associated with this commit. This is the file modification time
+    // from the commit's [`FileMeta`].
+    //
+    //
+    // TODO when incommit timestamps are supported: If there is a [`CommitInfo`] with a timestamp
+    // generated by in-commit timestamps, that timestamp will be used instead.
+    //
+    // Note: This will be used once an expression is introduced to transform the engine data in
+    // [`TableChangesScanData`]
+    #[allow(unused)]
+    timestamp: i64,
+}
+
+impl LogReplayScanner {
+    /// Constructs a LogReplayScanner, performing the Prepare phase detailed in [`LogReplayScanner`].
+    /// This iterates over each action in the commit.
+    fn prepare_table_changes(
+        commit_file: ParsedLogPath,
+        engine: &dyn Engine,
+        table_schema: &SchemaRef,
+    ) -> DeltaResult<Self> {
+        let visitor_schema = PreparePhaseVisitor::schema()?;
+        let action_iter = engine.get_json_handler().read_json_files(
+            &[commit_file.location.clone()],
+            visitor_schema,
+            None,
+        )?;
```

Comment:
this is duplicated here and below in the `into_scan_batches`. shouldn't we do this once and just try to iterate over it twice? can we put as a new field in `LogReplayScanner`

Reply:
Iterators can only be iterated over once. Iterating once would mean we have to save every engine data in memory and iterate over it again.

Reply:
... and there is no upper bound on the number of actions in a commit. Shallow cloning or replacing a massive table could produce a commit with hundreds of millions of file actions. Which is why we allow the json reader to return an iterator of batches even for a single file of input.